source/scene/scenenode.ts

```diff
@@ -0,0 +1,147 @@
+
+
+import { mat4 } from 'gl-matrix';
+
+import { SceneNodeComponent } from './scenenodecomponent';
+
+
+/**
+ * @todo add description
+ */
+export class SceneNode {
+
+    /** @see {@link name} */
+    protected _name: string;
+
+    /** @see {@link parent} */
+    protected _parent: SceneNode | undefined;
+
+    /** @see {@link nodes} */
+    protected _nodes = new Array<SceneNode>();
+
+    /** @see {@link transform} */
+    protected _transform: mat4;
+
+    /** @see {@link components} */
+    protected _components = new Array<SceneNodeComponent>();
+
+
+    /**
+     * @todo comment
+     * @param parent - @todo comment
+     */
+    constructor(name: string) {
+        this._name = name;
+        this._parent = undefined;
+    }
+
+
+    /**
+     * @todo comment
+     */
+    traverse(callback: SceneNode.NodeCallback): void {
+        callback(this);
+
+        /* Forwarding traversal. */
+        for (const child of this._nodes) {
+            child.traverse(callback);
+        }
+    }
+
+
+    /**
+     * @todo comment
+     * @param node - @todo comment
+     */
+    addNode(node: SceneNode): SceneNode {
+        node._parent = this;
+        this._nodes.push(node);
+
+        return node;
+    }
+
+    /**
+     * @todo comment
+     * @param component - @todo comment
+     */
+    addComponent(component: SceneNodeComponent): SceneNodeComponent {
+        this._components.push(component);
+
+        return component;
+    }
+
+    /**
+     * @todo comment
+     * @param type - @todo comment
+     */
+    componentsOfType(type: string): Array<SceneNodeComponent> {
+        return this._components.filter((component: SceneNodeComponent) => type === component.type);
+    }
+
+
+    /**
+     * Read-only access to ... @todo comment
+     */
+    get name(): string {
+        return this._name;
+    }
+
+    /**
+     * Read-only access to ... @todo comment
+     */
+    get parent(): SceneNode | undefined {
+        return this._parent;
+    }
+
+    /**
+     * Read-only access to ... @todo comment
+     */
+    get nodes(): Array<SceneNode> | undefined {
+        return this._nodes;
+    }
+
+    /**
+     * Read-only access to ... @todo comment
+     */
+    get transform(): mat4 {
```

Comment:
i would prefer to handle transformations as components as well, since this is already a specialization that might not be required. I might use sceneNode as hierarchy description just for the sake of hierarchy, or for naming composition, layer extraction, or as hierarchy traversal utility. I have lots of scenarios that would benefit from a generic scenenode/component system without and were I would like to switch the existing impl. to use scene nodes. None of them, however, requires transform.

Reply:
we could also think about multi parent relation ships... currently there is only one. We should redesign this as early as possible.